frontend/app/components/session-timeout.tsx

```diff
@@ -0,0 +1,108 @@
+import { useEffect, useState } from 'react';
+
+import { useFetchers, useLocation, useNavigation } from 'react-router';
+
+import { useTranslation } from 'react-i18next';
+import type { IIdleTimerProps } from 'react-idle-timer';
+import { useIdleTimer } from 'react-idle-timer';
+
+import { Button } from '~/components/button';
+import { Dialog, DialogContent, DialogFooter, DialogHeader, DialogTitle } from '~/components/dialog';
+
+export interface SessionTimeoutProps extends Required<Pick<IIdleTimerProps, 'promptBeforeIdle' | 'timeout'>> {
+  /**
+   * Function to call when the session end is triggered.
+   */
+  onSessionEnd: () => Promise<void> | void;
+
+  /**
+   * Function to call when the session extend is triggered.
+   */
+  onSessionExtend: () => Promise<void> | void;
+}
+
+/**
+ * SessionTimeout component to handle user session timeout prompts.
+ * This component uses the `react-IdleTimer` library to manage idle time and prompt the user
+ * before the session expires. It integrates with Remix's to activate the IdleTimer on route changes,
+ * fetcher submissions, and form submissions.
+ */
+export function SessionTimeout({ promptBeforeIdle, timeout, onSessionEnd, onSessionExtend }: SessionTimeoutProps) {
+  const { t } = useTranslation(['gcweb']);
+  const [timeRemaining, setTimeRemaining] = useState('');
+
+  const { activate, isPrompted, getRemainingTime } = useIdleTimer({
+    // Disable default event listeners; The IdleTimer should only activate during route navigation and form
+    // submissions, as these actions will interact with the session and extend its lifespan.
+    events: [],
+    onIdle: onSessionEnd,
+    promptBeforeIdle,
+    timeout,
+  });
+
+  // Track the current location and its key for detecting navigation changes.
+  const { key: locationKey } = useLocation();
+
+  useEffect(() => {
+    // Activate the IdleTimer whenever the location changes (indicating a route navigation).
+    activate();
+  }, [locationKey, activate]);
+
+  // Track fetcher states to activate the IdleTimer during submissions.
+  const fetchers = useFetchers();
+  const fetcherSubmitting = fetchers.filter(({ state }) => state === 'submitting').length > 0;
```

Comment:
Use `array.some()` funciton.

```suggestion
  const fetcherSubmitting = fetchers.some(({ state }) => state === 'submitting');
```